### PR TITLE
Add AutocompleteContext for autocomplete callbacks

### PR DESCRIPTION
**Pull request by zeffo**
## Summary

This PR adds AutocompleteContext. This context is passed to `Option.autocomplete`.

## Checklist

- [X] If code changes were made then they have been tested.
    - [X] I have updated the documentation to reflect the changes.
- [ ] This PR fixes an issue.
- [X] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, typehinting, ...)

### CODE REVIEW

File: discord/commands/context.py [hunk 1]
```diff
@@ -142,3 +142,32 @@ def cog(self) -> Optional[Cog]:
             return None
        
         return self.command.cog
+
+
+class AutocompleteContext:
+    """Represents a Discord autocomplete context.
```
Comment (nerdguyahmad) [inline discord/commands/context.py:148]:
```suggestion
    """Represents context for a slash command's option autocomplete.
```

File: discord/commands/context.py [hunk 2]
```diff
@@ -142,3 +142,32 @@ def cog(self) -> Optional[Cog]:
...
+    interaction: :class:`.Interaction`
+        The interaction object that invoked the autocomplete.
+    command: :class:`.ApplicationCommand`
+        The command that this context belongs to.
+    focused: :class:`.Option`
+        The option the user is currently typing.
+    value: :class:`.str`
+        The content of the focused option.
+    options :class:`.dict`
+        The names and values of all the options.
```
Comment (nerdguyahmad) [inline discord/commands/context.py:165]:
```suggestion
        A name to value mapping of the options that user has typed before this option.
```

**Review by nerdguyahmad — CHANGES_REQUESTED** [review]
Docs fixes, Ignore the syntax highlighting in suggestions. they seem broken or i don't know how to use them.

File: discord/commands/context.py (hunk 2 above)
Comment (Lulalaby) [reply discord/commands/context.py:165]:
```suggestion
        A name to value mapping of the options that the user has selected before this option.
```

**Review by Lulalaby — CHANGES_REQUESTED** [review]

**Comment by zeffo** [comment]
This should be slotted, no?

**Comment by nerdguyahmad** [comment]
> This should be slotted, no?

Yes, it should be.

File: discord/commands/context.py [hunk 3]
```diff
@@ -142,3 +142,32 @@ def cog(self) -> Optional[Cog]:
...
+    options :class:`.dict`
+        A name to value mapping of the options that the user has selected before this option.
+    """
+
+    def __init__(self, interaction: Interaction, *, command: ApplicationCommand, focused: Option, value: str, options: dict) -> None:
+        self.interaction = interaction
+        self.command = command
+        self.focused = focused
+        self.value = value
+        self.options = options
```
Comment (Middledot) [inline discord/commands/context.py:173]:
```suggestion
        self.options = options

```

**Comment by zeffo** [comment]
fuck i messed something up

**Comment by zeffo** [comment]
can someone do black git magic and fix my shit, ty

**Comment by nerdguyahmad** [comment]
Everything's fine.

**Comment by zeffo** [comment]
image
Are these supposed to be here

**Comment by Lulalaby** [comment]
> image
> Are these supposed to be here

Yes because I committed it

**Review by nerdguyahmad — DISMISSED** [review]

**Review by Lulalaby — APPROVED** [review]
LGTM

File: discord/commands/context.py [hunk 4]
```diff
@@ -142,3 +142,34 @@ def cog(self) -> Optional[Cog]:
...
+    """
+
+    __slots__ = ("interaction", "command", "focused", "value", "options")
+    
+    def __init__(self, interaction: Interaction, *, command: ApplicationCommand, focused: Option, value: str, options: dict) -> None:
+        self.interaction = interaction
+        self.command = command
+        self.focused = focused
+        self.value = value
+        self.options = options
```
Comment (BobDotCom) [inline discord/commands/context.py:175]:
```suggestion
        self.options = options

```

**Review by BobDotCom — CHANGES_REQUESTED** [review]
Please update utils.basic_autocomplete for this functionality

**Review by CodeWithSwastik — APPROVED** [review]

**Pull request merged**